p2p/transport_mconn.go

```diff
@@ -0,0 +1,646 @@
+	if m.listener != nil {
+		return errors.New("MConn transport is already listening")
+	}
+	err := m.normalizeEndpoint(&endpoint)
+	if err != nil {
+		return fmt.Errorf("invalid MConn listen endpoint %q: %w", endpoint, err)
+	}
+
+	m.listener, err = net.Listen("tcp", fmt.Sprintf("%v:%v", endpoint.IP, endpoint.Port))
+	if err != nil {
+		return err
+	}
+	if m.maxIncomingConnections > 0 {
+		m.listener = netutil.LimitListener(m.listener, m.maxIncomingConnections)
+	}
+
+	// Spawn a goroutine to accept inbound connections asynchronously.
+	go m.accept()
+
+	return nil
+}
+
+// accept accepts inbound connections in a loop, and asynchronously handshakes
+// with the peer to avoid head-of-line blocking. Established connections are
+// passed to Accept() via the channel m.chAccept.
+// See: https://github.com/tendermint/tendermint/issues/204
+func (m *MConnTransport) accept() {
+	for {
+		tcpConn, err := m.listener.Accept()
+		if err != nil {
+			select {
+			case m.chError <- err:
+			case <-m.chClose:
+			}
+			return
+		}
+		go func() {
+			err := m.filterTCPConn(tcpConn)
+			if err != nil {
+				_ = tcpConn.Close()
+				select {
+				case m.chError <- err:
+				case <-m.chClose:
+				}
+			}
+			conn, err := newMConnConnection(m, tcpConn, "")
+			if err != nil {
+				m.conns.Remove(tcpConn)
+				_ = tcpConn.Close()
+				select {
+				case m.chError <- err:
+				case <-m.chClose:
+				}
+			} else {
+				select {
+				case m.chAccept <- conn:
+				case <-m.chClose:
+					_ = conn.Close()
+				}
+			}
+		}()
+	}
+}
+
+// Accept implements Transport.
+func (m *MConnTransport) Accept(ctx context.Context) (Connection, error) {
+	select {
+	case conn := <-m.chAccept:
+		return conn, nil
+	case err := <-m.chError:
+		return nil, err
+	case <-m.chClose:
+		return nil, ErrTransportClosed{}
+	case <-ctx.Done():
+		return nil, nil
+	}
+}
+
+// Dial implements Transport.
+func (m *MConnTransport) Dial(ctx context.Context, endpoint Endpoint) (Connection, error) {
+	err := m.normalizeEndpoint(&endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, m.dialTimeout)
+	defer cancel()
+	dialer := net.Dialer{}
+	tcpConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%v:%v", endpoint.IP, endpoint.Port))
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.filterTCPConn(tcpConn)
+	if err != nil {
+		_ = tcpConn.Close()
+		return nil, err
+	}
+
+	conn, err := newMConnConnection(m, tcpConn, endpoint.PeerID)
+	if err != nil {
+		m.conns.Remove(tcpConn)
+		_ = tcpConn.Close()
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// Endpoints implements Transport.
+func (m *MConnTransport) Endpoints() []Endpoint {
+	if m.listener == nil {
+		return []Endpoint{}
+	}
+	addr := m.listener.Addr().(*net.TCPAddr)
+	return []Endpoint{{
+		Protocol: MConnProtocol,
+		PeerID:   m.nodeInfo.ID(),
+		IP:       addr.IP,
+		Port:     uint16(addr.Port),
+	}}
+}
+
+// Close implements Transport.
+func (m *MConnTransport) Close() error {
+	var err error
```

Comment:
We could just use a named return here to save a line of code and since the method is pretty concise.

Reply:
Yeah, I try to avoid named return values, I think they can be a bit easy to miss and get wrong.

Reply:
I do as well! I only use them when there typically is a `defer` used _and_ the function/method is very concise.